app/gui/suggestion-database/src/lib.rs

```diff
@@ -476,22 +480,29 @@ impl SuggestionDatabase {
 
     /// Search the database for an entry at `fully_qualified_name`. The parameter is expected to be
     /// composed of segments separated by the [`ACCESS`] character.
-    pub fn lookup_by_qualified_name_str(&self, qualified_name_str: &str) -> Option<Rc<Entry>> {
-        let qualified_name = QualifiedName::from_text(qualified_name_str).ok()?;
+    pub fn lookup_by_qualified_name_str(
+        &self,
+        qualified_name_str: &str,
+    ) -> FallibleResult<Rc<Entry>> {
+        let qualified_name = QualifiedName::from_text(qualified_name_str)?;
         let (_, entry) = self.lookup_by_qualified_name(&qualified_name)?;
-        Some(entry)
+        Ok(entry)
     }
 
     /// Search the database for an entry at `name` consisting fully qualified name segments, e.g.
     /// [`model::QualifiedName`].
     pub fn lookup_by_qualified_name<'a>(
         &self,
         name: impl Into<QualifiedNameRef<'a>>,
-    ) -> Option<(SuggestionId, Rc<Entry>)> {
+    ) -> FallibleResult<(SuggestionId, Rc<Entry>)> {
         let name = name.into();
         let segments = name.segments().map(CloneRef::clone_ref);
-        let id = self.qualified_name_to_id_map.borrow().get(segments);
-        id.and_then(|id| Some((id, self.lookup(id).ok()?)))
+        let id = self
+            .qualified_name_to_id_map
+            .borrow()
+            .get(segments)
+            .ok_or(NoSuchEntryWithName(name.to_string()))?;
```

Comment:
This eagerly allocates the string even if the lookup succeeded. You could use `ok_or_else` to prevent that.